include/pando-lib-galois/graphs/mirror_dist_local_csr.hpp

```diff
@@ -0,0 +1,482 @@
+    for (std::uint64_t i = 0; i < numHosts; i++) {
+      pando::Place place = pando::Place{pando::NodeIndex{static_cast<std::int16_t>(i)},
+                                        pando::anyPod, pando::anyCore};
+      PANDO_CHECK(pando::executeOn(place, mirrorAttach, vertexData, mirrorList, i, wgh));
+      local_mirror_size = lift(mirrorList[i], size);
+      numVertices += local_mirror_size;
+      _mirror_size += local_mirror_size;
+    }
+    PANDO_CHECK(wg.wait());
+    wgh.add(numHosts);
+
+    this->dlcsr.initializeAfterGather(vertexData, numVertices, edgeData, edgeMap, numEdges,
+                                      virtualToPhysical);
+
+    // Generate masterRange, mirrorRange, localMirrorToRemoteMasterOrderedTable
+    auto generateMetadata = +[](MirrorDistLocalCSR<VertexType, EdgeType> mdlcsr,
+                                DistLocalCSR<VertexType, EdgeType> dlcsr,
+                                HostLocalStorage<pando::Array<std::uint64_t>> mirrorList,
+                                std::uint64_t i, galois::WaitGroup::HandleType wgh) {
+      pando::Array<std::uint64_t> localMirrorList = mirrorList[i];
+      uint64_t mirror_size = lift(localMirrorList, size);
+      CSR csrCurr = dlcsr.arrayOfCSRs[i];
+
+      LocalVertexRange _masterRange = mdlcsr.masterRange.getLocalRef();
+      _masterRange = LocalVertexRange(lift(csrCurr, vertexEdgeOffsets.begin),
+                                      lift(csrCurr, size) - mirror_size);
+
+      LocalVertexRange _mirrorRange = mdlcsr.mirrorRange.getLocalRef();
+      _mirrorRange = LocalVertexRange(
+          lift(csrCurr, vertexEdgeOffsets.begin) + lift(csrCurr, size) - mirror_size, mirror_size);
+
+      pando::Array<MirrorToMasterMap> _localMirrorToRemoteMasterOrderedTable =
+          mdlcsr.localMirrorToRemoteMasterOrderedTable.getLocalRef();
+      fmap(_localMirrorToRemoteMasterOrderedTable, initialize, mirror_size);
+      for (uint64_t j = 0; j < mirror_size; j++) {
+        _localMirrorToRemoteMasterOrderedTable[j] =
+            MirrorToMasterMap(fmap(dlcsr, getLocalTopologyID, localMirrorList[j]),
+                              fmap(dlcsr, getGlobalTopologyID, localMirrorList[j]));
+      }
+      mdlcsr.masterRange.getLocalRef() = _masterRange;
+      mdlcsr.mirrorRange.getLocalRef() = _mirrorRange;
+      mdlcsr.localMirrorToRemoteMasterOrderedTable.getLocalRef() =
+          _localMirrorToRemoteMasterOrderedTable;
+      wgh.done();
+    };
+
+    for (std::uint64_t i = 0; i < numHosts; i++) {
+      pando::Place place = pando::Place{pando::NodeIndex{static_cast<std::int16_t>(i)},
+                                        pando::anyPod, pando::anyCore};
+      PANDO_CHECK(
+          pando::executeOn(place, generateMetadata, *this, this->dlcsr, mirrorList, i, wgh));
+      numVertices += lift(mirrorList[i], size);
+    }
+    PANDO_CHECK(wg.wait());
+
+    PANDO_CHECK_RETURN(setupCommunication());
+
+    return pando::Status::Success;
+  }
```

Comment:
Is this PR supposed to include all of this work?

Reply:
No, but the progress on MDLCSR construction PR is too slow so I have to base on the old version and work on it.
Do you want me to wait until the PR for MDLCSR construction to pass?

Reply:
So, you could branch out from current MDLCSR construction, not the main branch. It just compiled, worked for most of the cases including all test cases. I'm not sure what's the problem....

Reply:
Use this as your base, 
https://github.com/utcs-scea/pando-mono/pull/17

Reply:
Use https://github.com/utcs-scea/pando-mono/pull/29 instead.